src/collapse/xdebug.rs

```diff
@@ -0,0 +1,367 @@
+                    RawEntryMut::Vacant(vacant) => {
+                        // `Box<str>` has same hash as `str`.
+                        vacant.insert_hashed_nocheck(hash, Box::from(current), duration);
+                    }
+                }
+            }
+
+            if is_exit {
+                current_stack.pop();
+            } else {
+                let _ = parts.next();
+                let func_name = parts.next();
+                let _ = parts.next();
+                let path_name = parts.next();
+
+                if let (Some(func_name), Some(path_name)) = (func_name, path_name) {
+                    current_stack.call(func_name, path_name);
+                }
+            }
+
+            prev_start_time = time;
+        }
+
+        for (key, value) in stacks {
+            writeln!(writer, "{} {}", current_stack.frames(&key), value)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl CallStack {
+    /// Create a function name interning call stack tracker.
+    ///
+    /// Popoulated with the constant builtins for inclusion, to enable a faster comparison.
+    pub fn new() -> Self {
+        CallStack {
+            strings: CALLS
+                .iter()
+                .enumerate()
+                .map(|(idx, name)| (name.to_owned().into(), idx))
+                .collect(),
+            interned_string: CALLS.iter().cloned().map(Rc::from).collect(),
+            calls: HashMap::new(),
+            interned: Vec::new(),
+            stack: Vec::with_capacity(16),
+        }
+    }
+
+    /// Push a `call` line on top of the stack.
+    pub fn call(&mut self, name: &str, path: &str) {
+        let new_or_not = match self.intern_str(name) {
+            Interned::Old(st @ Str(0..=4)) => match self.intern_str(path) {
+                Interned::Old(other) => Interned::Old(Call::WithPath(st, other)),
+                Interned::New(new) => Interned::New(Call::WithPath(st, new)),
+            },
+            Interned::Old(other) => Interned::Old(Call::WithoutPath(other)),
+            Interned::New(new) => Interned::New(Call::WithoutPath(new)),
+        };
+
+        let idx = self.intern(new_or_not);
+        self.stack.push(idx)
+    }
+
+    /// Pop one call from the current stack.
+    pub fn pop(&mut self) {
+        self.stack.pop();
+    }
+
+    /// An empty stack means we finished main.
+    fn is_empty(&self) -> bool {
+        self.stack.is_empty()
+    }
+
+    /// Get a comparable, hashable representation of the call stack. of the call stack.
+    fn current(&self) -> &[usize] {
+        self.stack.as_slice()
+    }
+
+    /// Intern a string, return the unique index.
+    ///
+    /// `Ok` when the string was already present.
+    fn intern_str(&mut self, string: &str) -> Interned<Str> {
+        let mut hasher = self.strings.hasher().build_hasher();
+        string.hash(&mut hasher);
+        let hash = hasher.finish();
+
+        let entry = self
```

Comment:
Given that you also do this trick elsewhere, I think it'd be useful to pull it out into a standalone function. Maybe what we really want here is a version of entry that takes a `Cow` now that I think about it...